File: contracts/python/oath_contracts.py
import base64
import copy
import json
import hashlib
from typing import Any

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey


def _reject_floats(value: Any) -> None:
    if isinstance(value, float):
        raise TypeError("oath-json-v1 accepts only JSON integers")
    if isinstance(value, list):
        for item in value:
            _reject_floats(item)
    elif isinstance(value, dict):
        for item in value.values():
            _reject_floats(item)


def canonical_json(value: Any) -> bytes:
    _reject_floats(value)
    return json.dumps(
        value,
        ensure_ascii=False,
        allow_nan=False,
        separators=(",", ":"),
        sort_keys=True,
    ).encode("utf-8")


def verify_signed_document(document: dict[str, Any]) -> bool:
    detached = document.get("signature")
    if not isinstance(detached, dict):
        return False
    if detached.get("algorithm") != "ed25519":
        return False
    try:
        public_key = base64.b64decode(detached["public_key"], validate=True)
        signature = base64.b64decode(detached["signature"], validate=True)
        if len(public_key) != 32 or len(signature) != 64:
            return False
        payload = copy.deepcopy(document)
        payload["signature"] = None
        signed = canonical_json(payload)
        if detached.get("canonicalization") == "oath-json-v1+oath-domain-sha256-v1":
            domain = detached.get("domain")
            if not isinstance(domain, str) or not domain:
                return False
            domain_bytes = domain.encode("utf-8")
            signed = hashlib.sha256(
                b"oath-domain-signature-v1\0"
                + len(domain_bytes).to_bytes(8, "big")
                + domain_bytes
                + hashlib.sha256(signed).digest()
            ).digest()
        elif detached.get("canonicalization") != "oath-json-v1" or "domain" in detached:
            return False
        Ed25519PublicKey.from_public_bytes(public_key).verify(signature, signed)
        return True
    except (InvalidSignature, KeyError, TypeError, ValueError):
        return False

File: contracts/javascript/oath-contracts.mjs
import { createHash, createPublicKey, verify } from "node:crypto";

const ED25519_SPKI_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

export function canonicalJson(value) {
  if (value === null) return "null";
  if (typeof value === "string") return JSON.stringify(value);
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value)) {
      throw new TypeError("oath-json-v1 accepts only safe JSON integers");
    }
    return String(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map(canonicalJson).join(",")}]`;
  }
  if (typeof value === "object") {
    const members = Object.keys(value)
      .sort((left, right) => Buffer.compare(Buffer.from(left), Buffer.from(right)))
      .map((key) => `${JSON.stringify(key)}:${canonicalJson(value[key])}`);
    return `{${members.join(",")}}`;
  }
  throw new TypeError(`unsupported oath-json-v1 value: ${typeof value}`);
}

function decodeBase64(value, expectedBytes) {
  if (typeof value !== "string") return null;
  const decoded = Buffer.from(value, "base64");
  if (decoded.length !== expectedBytes || decoded.toString("base64") !== value) return null;
  return decoded;
}

export function verifySignedDocument(document) {
  const detached = document?.signature;
  if (detached?.algorithm !== "ed25519") return false;

  const publicKey = decodeBase64(detached.public_key, 32);
  const signature = decodeBase64(detached.signature, 64);
  if (publicKey === null || signature === null) return false;

  const payload = { ...document, signature: null };
  let signed = Buffer.from(canonicalJson(payload));
  if (detached?.canonicalization === "oath-json-v1+oath-domain-sha256-v1") {
    if (typeof detached.domain !== "string" || detached.domain.length === 0) return false;
    const domain = Buffer.from(detached.domain);
    const length = Buffer.alloc(8);
    length.writeBigUInt64BE(BigInt(domain.length));
    signed = createHash("sha256")
      .update(Buffer.from("oath-domain-signature-v1\0"))
      .update(length)
      .update(domain)
      .update(createHash("sha256").update(signed).digest())
      .digest();
  } else if (detached?.canonicalization !== "oath-json-v1" || detached.domain !== undefined) {
    return false;
  }
  const key = createPublicKey({
    key: Buffer.concat([ED25519_SPKI_PREFIX, publicKey]),
    format: "der",
    type: "spki",
  });
  return verify(null, signed, key, signature);
}
